### PR TITLE
Support for linear and radial gradients.

### PR DESCRIPTION
**Pull request by stevesheldon**
@articulate/mac - please review.

### CODE REVIEW

File: source/IJSVGColor.m [hunk 1]
```diff
@@ -616,10 +616,8 @@ + (NSColor *)colorFromHEXString:(NSString *)string
         NSInteger r = (hex>>16) & 0xFF;
         NSInteger g = (hex>>8) & 0xFF;
         NSInteger b = (hex) & 0xFF;
-        return [NSColor colorWithDeviceRed:r/255.f
-                                     green:g/255.f
-                                      blue:b/255.f
-                                     alpha:alpha];
+
+        return [NSColor colorWithRed255:r green255:g blue255:b alpha255:255.0 * alpha];
```
Comment (stevesheldon) [inline source/IJSVGColor.m:620]: The colors being set were not correct, they need to map to the SRGB color space from web values.

File: source/IJSVGGradientLayer.m [hunk 2]
```diff
@@ -34,9 +36,171 @@ - (void)drawInContext:(CGContextRef)ctx
...
+            
+            endPoint.x = CGRectGetWidth(objectRect)*endPoint.x;
+            endPoint.y = CGRectGetHeight(objectRect)*endPoint.y;
+            
+        } else {
+            
+            // apply the transforms specified on the gradient element
+            for(IJSVGTransform * gradientTransform in self.gradient.transforms) {
+                applyTransform(gradientTransform.CGAffineTransform);
+            }
```
Comment (stevesheldon) [inline source/IJSVGGradientLayer.m:115]: Most of the code is from the merge, I added this transform to adhere to the spec for SVG.

File: source/IJSVGShapeLayer.m [hunk 3]
```diff
@@ -17,6 +19,46 @@ - (void)dealloc
...
+        [super renderInContext:ctx];
+        CGContextRestoreGState(ctx);
+        return;
+    }
+    [super renderInContext:ctx];
+}
+
+- (CGPoint)absoluteOrigin
+{
+    return CGPointZero;
```
Comment (stevesheldon) [inline source/IJSVGShapeLayer.m:61]: I found this by accident but the absolute origin for shapes should always be zero, if we use the calculated value for this particular kind of layer, gradients don't render in the correct location within the mask.

**Comment by rthompson0702** [comment]
Looks good to me.  I don't know much about the math and transformations but the results look good. 👍

**Pull request merged**